### ISSUE TITLE
Marshal/MarshalTo with value receiver

### ISSUE THREAD

**Issue by anton-povarov** [gogo/protobuf#300]
TL/DR: Let's generate copies of Marshal/MarshalTo methods with value receivers, this should help reduce GC pressure for messages with optional (aka pointer) fields

Consider a message like this
```protobuf
message response_generic {
	required sint32 error_code = 1;
	optional string error_text = 2;
}
```

enabling unmarshaler, code like this will be generated
```go
type ResponseGeneric struct {
	ErrorCode *int32  `protobuf:"zigzag32,1,req,name=error_code" json:"error_code,omitempty"`
	ErrorText *string `protobuf:"bytes,2,opt,name=error_text" json:"error_text,omitempty"`
}

func (m *ResponseGeneric) Size() (n int) {...}
func (m *ResponseGeneric) Marshal() (data []byte, err error) {...}
func (m *ResponseGeneric) MarshalTo(data []byte) (int, error) {...}
```

now consider this benchmark, it gives us 3 allocs per op (fields escape + the resulting buffer i assume)
```go
func prepare___Ptr() *ResponseGeneric {
	res := ResponseGeneric{
		ErrorCode: proto.Int32(0),
		ErrorText: proto.String("long and annoying string"),
	}
	return &res
}
func BenchmarkMarshal__Ptr(b *testing.B) {

	for i := 0; i < b.N; i++ {
		msg := prepare___Ptr()
		_, _ = msg.Marshal()
	}
}
```

```sh
BenchmarkMarshal__Ptr-4          5000000               270 ns/op             148 B/op          3 allocs/op
```

and here comes the interesting part, suppose we add two methods, that are completely identical to `Marshal` and `MarshalTo`, but take their argument by value, not by pointer
```go
func (m ResponseGeneric) MarshalValue() (data []byte, err error) {...}
func (m ResponseGeneric) MarshalValueTo(data []byte) (int, error) {...}
```

and a benchmark
```go
func prepare___Value() ResponseGeneric {
	res := ResponseGeneric{
		ErrorCode: proto.Int32(0),
		ErrorText: proto.String("long and annoying string"),
	}
	return res
}

func BenchmarkMarshal__Value(b *testing.B) {

	for i := 0; i < b.N; i++ {
		msg := prepare___Value()
		_, _ = msg.MarshalValue()
	}
}
```

and now values do not escape
```sh
BenchmarkMarshal__Value-4       10000000               169 ns/op             128 B/op          1 allocs/op
```

full code: https://gist.github.com/anton-povarov/40fcd49c27f0d721339c3b9b0591d0c3

**Comment by anton-povarov** [comment]
created a patch to kick off the discussion
https://gist.github.com/anton-povarov/536d529c50159b13dc6f2ab2c264786a

**Comment by awalterschulze** [comment]
Hey welcome back :)  Long time no see.
I suspect this is only true for small messages, since then the copy of the receiver to the Marshal "function" is small.
What happens when we have lots of fields and/or the fields are not pointers?

**Comment by anton-povarov** [comment]
Hey hey :)

Well, it's not that i'm after cpu nanoseconds in this case, trying to reduce allocations.

Testing with a message, 47 fields (all optional = pointers, integers/floats/strings, no embedded messages).
Set all of them, to simulate "not-nullable".

It's still significantly faster to copy
```sh
BenchmarkMarshal__UserPtrInlined-4               1000000              1396 ns/op             288 B/op         29 allocs/op
BenchmarkMarshal__UserValueInlined-4             3000000               559 ns/op             128 B/op          1 allocs/op
```

But, unfortunately, it seems that my goal of reducing allocations can't be reached by changes i gogoprotobuf alone (although, i'd still like to see them implemented :) ).

The culprit seems to be with a combo of escape analysis and function inlining. If message is returned from a function (even by value), but that function is not inlined - sub-values are still going to escape and ruin the whole thing ;(

**Comment by anton-povarov** [comment]
too many hacks are required (aka, spaghetti code) to take full advantage of this at the moment :(
this change to 1.9 should help i guess: https://github.com/golang/go/issues/19348

**Comment by awalterschulze** [comment]
Wow lots of interesting things here.

What happens with non pointer fields?
I don't think populating the value counts as simulating non nullable.

**Comment by anton-povarov** [comment]
You mean when fields are non-pointer, the benchmark would not need to allocate and becomes faster? Sure it would, but that's not the case we're looking at :)

I guess i could make a full example with compileable protofile, just adding more fields to the message.

Thing is - if a function like `prepare___Value` from the OP is not inlined - then pointer fields will be heap allocated (since escape analysis is... not ideal). And this will happen with messages large enough.

And the way to avoid that - is to BOTH inline the function manually and generate+call the `MarshalValueTo` method.

With that done - one could also reuse buffers, passed to `MarshalValueTo` and achieve very low allocation rate.

**Comment by anton-povarov** [comment]
Stay tuned for full example.

p.s. have you explored the idea of implementing interface similar to native C++ generator, aka generating has_ flags and setters ?

**Comment by awalterschulze** [comment]
I have not explored the has_flags and setters idea.  I have thought about it though.  I decided on the raw struct as this is the way that my first users preferred to work.  And then this whole project was easier to sell.  And if the struct that is generated is not the one you want to use, then there is a copy from the generated struct to the one you want to use.  This is super expensive and hard to maintain.
But I am more open to Setters these days, since it could then enable the use of private fields.
This is a separate issue from the has_flags field, but it could be considered a step in that direction.

**Comment by awalterschulze** [comment]
Ok so when its inlined, then there will be no copy overhead and then the amount of non-nullable fields won't make a difference.  I was worried that given enough fields the copy of the receiver to the function will become quite large.  But I seemed to have missed the whole point of what you are doing here.

**Comment by awalterschulze** [comment]
Ok I am a bit slow today, sorry.
Let me try to summarize my understanding.

```
func (m *ResponseGeneric) Marshal() (data []byte, err error) {...}
```
requires that `m` be put on the stack, which results in an alloc.

```
func (m ResponseGeneric) MarshalValue() (data []byte, err error) {...}
```
requires that `m` be copied to the function.
But if the function is inlined that `m` is not copied and you are winning.

**Comment by awalterschulze** [comment]
Is this right?

**Comment by anton-povarov** [comment]
Um, as far as i understand - no :)

Basically, when escape analysis decides where to put m, it has to consider if there is any code, that might use `m`, after function that created `m` exits.

Given
```go
msg := prepare___TestPtr()
_, _ = msg.Marshal()
```

Escape analysis has to decide (broadly), if there can ever be any pointer to `msg` that outlives current goroutine. and if yes - it has to be allocated on heap. Full program analysis is required to discover that, and it's slow. So Go tries it's best, but not doesn't analyze too 'deep'.

From my incomplete understanding - it follows 2 levels of indirection deep only.
1st pointer is - msg itself, coming from the function.
2nd pointer is - the field itself

and then there is a function call with pointer arg - that's the 3rd ptr i assume. So escape then gives up, and just puts everything to heap, just to be safe.

What we can do - is remove 'levels of indirection' though pointers, passing values to be copied.
This does add overhead (need to copy to `MarshalValue`), but that can be optimised by the compiler.
But removing those levels of indirection - we help escape analysis realise - that all fields even if pointer) of `m` can be safely stored on stack, since no function (i.e. `MarshalValue`) saves any pointers to those fields or the message itself.

**Comment by anton-povarov** [comment]
full code, need to apply patch from OP to compile.
https://gist.github.com/anton-povarov/1905e64b9c2d66d4d44724c8ff441eee

the more fields you add - slower value methods are. non-nullable fields are always faster of course.

```sh
BenchmarkMarshal__TestPtr-4                      2000000               779 ns/op             168 B/op         14 allocs/op
BenchmarkMarshal__TestValue-4                    5000000               309 ns/op              64 B/op          1 allocs/op
BenchmarkMarshal__TestNotNullablePtr-4           2000000               667 ns/op             192 B/op          1 allocs/op
BenchmarkMarshal__TestNotNullableValue-4         2000000               730 ns/op             192 B/op          1 allocs/op
```

**Comment by anton-povarov** [comment]
Fun effect i didn't think of, non-nullable messages incur less allocs, but more bytes allocated, since they need to store full string headers in struct.
and pointer fields can just use pointers to interned strings.

doesn't matter much though.

**Comment by anton-povarov** [comment]
After some more testing i'll have to concede defeat for now.
Need this to work with slices of pb messages, and it just doesn't, will have to find some other way.

Sorry for the inconvenience.

**Comment by anton-povarov** [comment]
Just a quick question, have you considered the idea of adding `(gogoproto.null_value)="some_value"` in addition to `(gogoproto.nullable)=false`?

i.e. some value shall be reserved for 'do not send this optional field' semantics.
looks awfully close to protobuf built-in `[default=some_value]` though.

**Comment by awalterschulze** [comment]
No worries I think I learned something :)
By removing pointers we are also making it easier for go to do escape analysis.
But there will almost always be too many pointers, given slices and nested messages.

**Comment by awalterschulze** [comment]
No I have not, since it makes it incompatible with non go unmarshalers.
Also:
  - proto2 or at least proto2 in golang/protobuf was still sending defaults.
  - proto3 has removed default and now only supports "zero" values, which it does not send.
